thetis/utility.py

```diff
@@ -1896,3 +1896,88 @@ def get_total_depth(self, eta):
         else:
             total_h = self.bathymetry_2d
         return total_h
+
+
+class DepthIntegratedPoissonSolver(object):
+    """
+    Construct solvers for Poisson equation and subsequently updating velocities
+
+    Generic form:
+       2D: `div(grad(q^{n+1/2})) + inner(A, grad(q^{n+1/2})) + B*q^{n+1/2} = C`
+
+    :arg A, B and C: Known functions, constants or expressions
+    :type A: vector, B: scalar, C: scalar (3D: div terms). Valued :class:`Function`, `Constant`, or an expression
+    :arg q: Non-hydrostatic pressure
+    :type q: scalar function :class:`Function`
+    """
+
+    def __init__(self, solver):
+        rho_0 = physical_constants['rho0']
+        self.fields = solver.fields
+        self.depth = solver.depth
+        self.dt = solver.dt
+        self.bnd_functions = solver.bnd_functions
+
+        q_2d = self.fields.q_2d
+        fs_q = q_2d.function_space()
+        test_q = TestFunction(fs_q)
+        normal = FacetNormal(fs_q.mesh())
+        boundary_markers = fs_q.mesh().exterior_facets.unique_markers
+
+        uv_2d, elev_2d = self.fields.solution_2d.split()
+        w_2d = solver.w_2d
+        bath_2d = self.fields.bathymetry_2d
+        h_star = self.depth.get_total_depth(elev_2d)
+        w_b = -dot(uv_2d, grad(bath_2d))  # TODO account for bed movement
+
+        A = grad(elev_2d - bath_2d)/h_star
+        B = div(A) - 4./(h_star**2)
+        C = 2.*rho_0/self.dt*(div(uv_2d) + (w_2d - w_b)/(0.5*h_star))
```

Comment:
I think I commented on this before, but this looks very suspect: if `elev_2d` is linear, then the gradient is piecewise constant, so you can't take a further derivative of it

Reply:
Does that make sense that the further derivative of the piecewise constant is zero? And if it's indeed inappropriate, may I ask for your specific suggestion about further corection? In terms of current treatment, all present test cases are ok.

Reply:
Yes the further derivative would be zero indeed. You have to remember that all you are doing is writing finite element integrals over either the element or over element boundaries. It's up to you to write a consistent Finite Element discretisation. Inside the elements all derivatives are perfectly well defined, but if you would do `v*div(grad(u))*dx` you would *not* get a consistent discretisation for continuous `u` you have to integrate by parts and `inner(grad(v), grad(u))*dx` is *not* the same as `v*div(grad(u))*dx` - the difference being `dS`-terms. For a _discontinous_ `u` things get more complicated and you have to go into different DG schemes for elliptic terms.

In the case of the `-q_2d*div(A*test_q)*dx` term, I think it would be correct to write:

`dot(grad(q_2d), A*test_q)*dx` instead because `q_2d` is continuous.

For the `B*q_2d*test_q*dx` term, I think you have to split it up. Before discretisation we have:
```
B = div(A) - 4./(h_star**2) = div(grad(elev_2d-bath_2d))/h_star - dot(grad(elev_2d-bath_2d)/h_star**2, grad(h_star)) - 4/h_star**2
```
The first term of `B*q_2d*test_q*dx` has a standard div-grad (i.e. Laplacian) of a continuous `elev_2d-bath_2d` which we know we can discretise by integrating by parts (this is valid as long as the test function and the q_2d/h_star on the outside are also continuous):
```
-dot(grad(test_q*q_2d/h_star), grad(elev_2d-bath_2d)) * dx
```
The second and term term are fine as they are:
```
- (dot(grad(elev_2d-bath_2d), grad(h_star)) - 4)/h_star**2*q_2d*test_q*dx
```
For the `-C*test_q*dx`-term, there is an issue with the `div(uv_2d)` term as our `uv_2d` is always discontinuous. That term again can be rewritten with IBP, as:
```
2.*rho_0/self.dt*dot(uv_2d, grad(test_q))*dx
```
(note the minus-sign disappears).

Hope this makes some sense. Do check my maths, cause I'm just typing this out in github.

It might be that the difference between what I suggest and what you have in practice is small, but I _think_ what I suggest is consistent, i.e. it should converge to the continuous (i.e before spatial discretisation) one-layer PDE you claim to be solving. It would also be interesting to see how these changes affect the polynomial order you need: for instance with quadratic elevation (which you use in one of the examples, I think), div(grad(elev_2d)) is at least nonzero (still not a consistent discretisation).

Reply:
Cool. That makes sense. I will try to edit. Thanks!